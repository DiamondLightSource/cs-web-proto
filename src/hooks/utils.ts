import { PvState, CsState } from "../redux/csState";

export interface PvArrayResults {
  [pvName: string]: [PvState, string];
}

export function pvStateSelector(
  pvNames: string[],
  state: CsState
): PvArrayResults {
  const results: PvArrayResults = {};
  for (const pvName of pvNames) {
    results[pvName] = [
      state.valueCache[pvName],
      state.shortPvNameMap[pvName] || pvName
    ];
  }
  return results;
}

/* Used for preventing re-rendering if the results are equivalent. */
export function pvStateComparator(
  before: PvArrayResults,
  after: PvArrayResults
): boolean {
  if (Object.keys(before).length !== Object.keys(after).length) {
    return false;
  }
  for (const [pvName, [beforeVal, beforeShortPvName]] of Object.entries(
    before
  )) {
    const [afterVal, afterShortPvName] = after[pvName];
    // If the PV state hasn't changed in the store, we will receive the same
    // object when selecting that PV state.
    if (beforeVal !== afterVal || beforeShortPvName !== afterShortPvName) {
      return false;
    }
  }
  return true;
}
