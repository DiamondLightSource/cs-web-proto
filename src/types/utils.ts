export function filterUndefinedOut(input: {
  [index: string]: any;
}): { [index: string]: any } {
  const output: { [index: string]: any } = {};
  let key;

  for (key in input) {
    if (input.hasOwnProperty(key) && input[key] !== undefined) {
      output[key] = input[key];
    }
  }

  return output;
}
