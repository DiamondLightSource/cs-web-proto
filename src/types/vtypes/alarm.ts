export enum AlarmSeverity {
  NONE,
  MINOR,
  MAJOR,
  INVALID,
  UNDEFINED
}

export enum AlarmStatus {
  NONE,
  DEVICE,
  DRIVER,
  RECORD,
  DB,
  CONF,
  UNDEFINED,
  CLIENT
}

export interface AlarmProvider {
  getAlarm(): Alarm;
}

export abstract class Alarm {
  public abstract getSeverity(): AlarmSeverity;
  public abstract getStatus(): AlarmStatus;
  public abstract getName(): string;
}

export class IAlarm extends Alarm {
  private severity: AlarmSeverity;
  private status: AlarmStatus;
  private name: string;
  public constructor(
    severity: AlarmSeverity,
    status: AlarmStatus,
    name: string
  ) {
    super();
    this.severity = severity;
    this.status = status;
    this.name = name;
  }
  public getSeverity(): AlarmSeverity {
    return this.severity;
  }
  public getStatus(): AlarmStatus {
    return this.status;
  }
  public getName(): string {
    return this.name;
  }
}

export const alarm = (
  severity: AlarmSeverity,
  status: AlarmStatus,
  name: string
): Alarm => {
  return new IAlarm(severity, status, name);
};

export const isAlarmProvider = (object: any): object is AlarmProvider => {
  return "getAlarm" in object;
};

export const ALARM_NONE = alarm(AlarmSeverity.NONE, AlarmStatus.NONE, "");

export const alarmOf = (object: any): Alarm => {
  if (object && isAlarmProvider(object)) {
    return object.getAlarm();
  }
  return ALARM_NONE;
};
