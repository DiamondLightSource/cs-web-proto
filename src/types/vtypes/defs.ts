export interface Range {
  min: number;
  max: number;
}

export const RANGE_NONE: Range = {
  min: 0,
  max: 0
};
