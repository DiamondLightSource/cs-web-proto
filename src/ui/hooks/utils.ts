import { PvState, CsState } from "../../redux/csState";

export interface PvArrayResults {
  [pvName: string]: [PvState, string];
}

export function pvStateSelector(
  pvNames: string[],
  state: CsState
): PvArrayResults {
  const results: PvArrayResults = {};
  for (const pvName of pvNames) {
    const effectivePvName = state.effectivePvNameMap[pvName] || pvName;
    results[pvName] = [state.valueCache[effectivePvName], effectivePvName];
  }
  return results;
}

export function deviceSelector(
  deviceName: string,
  state: CsState
): {} {
  return state.deviceCache[deviceName];
}

/* Used for preventing re-rendering if the results are equivalent.
   Note that if the state for a particular PV hasn't changed, we will
   get back the same object as last time so we are safe to compare them.
   We need to be careful that we don't have a situation where we get back
   the same object with different properties and compare it as equal when
   in fact it has changed.
*/
export function pvStateComparator(
  before: PvArrayResults,
  after: PvArrayResults
): boolean {
  if (Object.keys(before).length !== Object.keys(after).length) {
    return false;
  }
  for (const [pvName, [beforeVal, beforeEffPvName]] of Object.entries(before)) {
    const [afterVal, afterEffPvName] = after[pvName];
    // If the PV state hasn't changed in the store, we will receive the same
    // object when selecting that PV state.
    if (beforeVal !== afterVal || beforeEffPvName !== afterEffPvName) {
      return false;
    }
  }
  return true;
}
  
  export function deviceComparator(
    before: {},
    after: {}
  ): boolean {
    if (Object.keys(before).length !== Object.keys(after).length) {
      return false;
    }
    return true;
  }
