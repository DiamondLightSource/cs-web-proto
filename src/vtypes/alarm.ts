export enum AlarmSeverity {
  NONE,
  MINOR,
  MAJOR,
  INVALID,
  UNDEFINED
}

export enum AlarmStatus {
  NONE,
  DEVICE,
  DRIVER,
  RECORD,
  DB,
  CONF,
  UNDEFINED,
  CLIENT
}

export abstract class Alarm {
  public abstract getSeverity(): AlarmSeverity;
  public abstract getStatus(): AlarmStatus;
  public abstract getName(): string;
}

class IAlarm extends Alarm {
  private severity: AlarmSeverity;
  private status: AlarmStatus;
  private name: string;
  public constructor(
    severity: AlarmSeverity,
    status: AlarmStatus,
    name: string
  ) {
    super();
    this.severity = severity;
    this.status = status;
    this.name = name;
  }
  public getSeverity(): AlarmSeverity {
    return this.severity;
  }
  public getStatus(): AlarmStatus {
    return this.status;
  }
  public getName(): string {
    return this.name;
  }
}

export const alarmOf = (
  severity: AlarmSeverity,
  status: AlarmStatus,
  name: string
): Alarm => {
  return new IAlarm(severity, status, name);
};

export const ALARM_NONE = alarmOf(AlarmSeverity.NONE, AlarmStatus.NONE, "");
