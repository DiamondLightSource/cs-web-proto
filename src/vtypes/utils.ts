import { VNumber, VType, vdoubleOf } from "./vtypes";
import { vstringOf } from "./string";
import { ALARM_NONE } from "./alarm";
import { timeNow } from "./time";
import { DISPLAY_NONE } from "./display";

export const vtypeToString = (vtype?: VType, precision?: number): string => {
  console.log("vtype");
  console.log(vtype);
  if (vtype instanceof VNumber) {
    if (precision) {
      return vtype.getValue().toFixed(precision);
    } else {
      return vtype.getValue().toString();
    }
  }
  if (vtype) {
    return vtype.toString();
  }
  return "";
};

export function vtypeToNumber(vtype: VType): number {
  let value = vtype.getValue();
  let numericValue;
  if (typeof value === "number") {
    numericValue = value;
  } else if (typeof value === "string") {
    numericValue = parseFloat(value);
  } else {
    numericValue = 0;
  }
  return numericValue;
}

export function vtypeOrUndefinedToNumber(vtype?: VType): number {
  if (vtype) {
    return vtypeToNumber(vtype);
  } else {
    return 0;
  }
}

export const stringToVtype = (
  value: string,
  alarm = ALARM_NONE,
  time = timeNow(),
  display = DISPLAY_NONE
): VType => {
  try {
    let numberValue = parseFloat(value);
    return vdoubleOf(numberValue, alarm, time, display);
  } catch (error) {
    return vstringOf(value, alarm, time);
  }
};

export const valueToVtype = (
  value: object,
  alarm = ALARM_NONE,
  time = timeNow(),
  display = DISPLAY_NONE
): VType => {
  if (typeof value === "string") {
    return vstringOf(value, alarm, time);
  } else if (typeof value === "number") {
    return vdoubleOf(value, alarm, time, display);
  }
  return vdoubleOf(0, alarm, time, display);
};
