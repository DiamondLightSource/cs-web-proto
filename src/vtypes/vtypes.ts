import { Alarm, AlarmProvider, ALARM_NONE } from "./alarm";
import { Display, DISPLAY_NONE } from "./display";
import { Time, timeNow, TimeProvider } from "./time";

export abstract class VType {
  public abstract getValue(): any;
}

export abstract class Scalar extends VType
  implements AlarmProvider, TimeProvider {
  public abstract getAlarm(): Alarm;
  public abstract getTime(): Time;
}

export abstract class Array extends VType {
  public abstract getSizes(): number[];
}

export abstract class VNumber extends Scalar {
  public abstract getValue(): number;
}

export type VNumberBuilder = (
  value: any,
  alarm?: Alarm,
  time?: Time,
  display?: Display
) => VType;

export type VNumberArrayBuilder = (
  value: any,
  sizes: number[],
  alarm?: Alarm,
  time?: Time,
  display?: Display
) => VType;

export abstract class VDouble extends VNumber {
  public abstract getValue(): number;
  public abstract getDisplay(): Display;
}

class IVDouble extends VDouble {
  private value: number;
  private alarm: Alarm;
  private time: Time;
  private display: Display;
  public constructor(
    value: number,
    alarm: Alarm,
    time: Time,
    display: Display
  ) {
    super();
    this.value = value;
    this.alarm = alarm;
    this.time = time;
    this.display = display;
  }
  public getValue(): number {
    return this.value;
  }
  public getAlarm(): Alarm {
    return this.alarm;
  }
  public getTime(): Time {
    return this.time;
  }
  public getDisplay(): Display {
    return this.display;
  }
}

export const vdouble = (
  double: number,
  alarm = ALARM_NONE,
  time = timeNow(),
  display = DISPLAY_NONE
): VDouble => new IVDouble(double, alarm, time, display);

export abstract class VNumberArray extends Array implements AlarmProvider {
  public abstract getValue(): number[];
  public abstract getTime(): Time;
  public abstract getAlarm(): Alarm;
  public abstract getDisplay(): Display;
}

export abstract class VDoubleArray extends VNumberArray {
  public abstract getValue(): number[];
}

class IVDoubleArray extends VDoubleArray {
  private value: number[];
  private sizes: number[];
  private alarm: Alarm;
  private time: Time;
  private display: Display;
  public constructor(
    value: number[],
    sizes: number[],
    alarm: Alarm,
    time: Time,
    display: Display
  ) {
    super();
    this.value = value;
    this.sizes = sizes;
    this.alarm = alarm;
    this.time = time;
    this.display = display;
  }
  public getValue(): number[] {
    return this.value;
  }
  public getSizes(): number[] {
    return this.sizes;
  }
  public getAlarm(): Alarm {
    return this.alarm;
  }
  public getTime(): Time {
    return this.time;
  }
  public getDisplay(): Display {
    return this.display;
  }
  public toString(): string {
    return this.value.toString();
  }
}

export const vdoubleArray = (
  value: number[],
  sizes: number[],
  alarm = ALARM_NONE,
  time = timeNow(),
  display = DISPLAY_NONE
): VDoubleArray => new IVDoubleArray(value, sizes, alarm, time, display);

abstract class EnumDisplay {
  public abstract getChoices(): string[];
}

class IEnumDisplay extends EnumDisplay {
  private choices: string[];

  public constructor(choices: string[]) {
    super();
    this.choices = choices;
  }

  public getChoices(): string[] {
    return this.choices;
  }
}

export abstract class VEnum extends Scalar {
  public abstract getValue(): string;
  public abstract getIndex(): number;
  public abstract getDisplay(): EnumDisplay;
}

class IVEnum extends VEnum {
  private index: number;
  private alarm: Alarm;
  private time: Time;
  private display: EnumDisplay;

  public constructor(
    index: number,
    display: EnumDisplay,
    alarm: Alarm,
    time: Time
  ) {
    super();
    this.index = index;
    this.alarm = alarm;
    this.time = time;
    this.display = display;
  }

  public getValue(): string {
    return this.display.getChoices()[this.index];
  }
  public getIndex(): number {
    return this.index;
  }
  public getDisplay(): EnumDisplay {
    return this.display;
  }
  public getAlarm(): Alarm {
    return this.alarm;
  }
  public getTime(): Time {
    return this.time;
  }
}

export const venum = (
  index: number,
  choices: string[],
  alarm: Alarm = ALARM_NONE,
  time: Time = timeNow()
): VEnum => new IVEnum(index, new IEnumDisplay(choices), alarm, time);
