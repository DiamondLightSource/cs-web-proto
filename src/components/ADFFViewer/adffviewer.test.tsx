import React from "react";
import {
  RawADFFViewerComponent,
  ADFFViewer,
  RawADFFViewerComponentPropsType,
  ADFFViewerWidgetPropsType
} from "./adffviewer";
import { configure, shallow, ShallowWrapper } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { stringToVtype } from "../../vtypes/utils";
import { create, ReactTestRenderer } from "react-test-renderer";

configure({ adapter: new Adapter() });

let wrapper: ShallowWrapper<ADFFViewerWidgetPropsType>;
let rawSnapshot: ReactTestRenderer;
let rawWrapper: ShallowWrapper<RawADFFViewerComponentPropsType>;
const FAKE_URL = "http://fakeurl.com/video.mjpeg";
const FAKE_PV_PREFIX = "pv-prefix-1";
const FAKE_PV = `${FAKE_PV_PREFIX}:MJPG_URL_RBV`;

beforeEach((): void => {
  const rawADFFViewerComponent = (
    <RawADFFViewerComponent
      connected={true}
      readonly={true}
      value={stringToVtype(FAKE_URL)}
    />
  );

  const adffviewer = (
    <ADFFViewer
      prefix={FAKE_PV_PREFIX}
      containerStyling={{
        position: "relative",
        width: 200,
        height: 200,
        margin: "0px",
        padding: "0px"
      }}
    />
  );

  rawWrapper = shallow(rawADFFViewerComponent);
  rawSnapshot = create(rawADFFViewerComponent);
  wrapper = shallow(adffviewer);
});

describe("<RawADFFViewerComponent />", (): void => {
  test("it matches the snapshot", (): void => {
    expect(rawSnapshot.toJSON()).toMatchSnapshot();
  });

  test("it renders a basic element", (): void => {
    // there is at least one img that could render the stream
    const img = rawWrapper.find("img");
    expect(img.length).toBeGreaterThan(0);
  });

  test("Element uses provided URL", (): void => {
    const img = rawWrapper.find("img");
    expect(img.props().src).toEqual(FAKE_URL);
  });
});

describe("<ADFFViewer />", (): void => {
  test("It derives the proper PV name", (): void => {
    expect(wrapper.find("RawADFFViewer").prop("pvName")).toEqual(FAKE_PV);
  });
});
