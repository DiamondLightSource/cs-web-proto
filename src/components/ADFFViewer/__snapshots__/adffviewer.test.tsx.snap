// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RawADFFViewerComponent /> it matches the snapshot 1`] = `
<div>
  <img
    alt=""
    src="http://fakeurl.com/video.mjpeg"
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  />
</div>
`;
